src/ui/GenioWindow.cpp

```diff
@@ -1683,99 +1683,129 @@ GenioWindow::_FileCloseAll()
+	}
+	if (fTabManager->CountTabs() > opened_index) {
+		fTabManager->SelectTab(opened_index);
+	}
+	return B_OK;
+}
 
 status_t
 GenioWindow::_FileOpen(BMessage* msg)
 {
-	int32 opened_index = msg->GetInt32("opened_index", fTabManager->SelectedTabIndex() + 1);
-
-	const int32 be_line   = msg->GetInt32("start:line", msg->GetInt32("be:line", -1));
-	const int32 lsp_char  = msg->GetInt32("start:character", -1);
-
-	BMessage selectTabInfo;
-	selectTabInfo.AddInt32("start:line", be_line);
-	selectTabInfo.AddInt32("start:character", lsp_char);
-
-	bool openWithPreferred	= msg->GetBool("openWithPreferred", false);
+	//start-up files.
+	if (msg->HasInt32("opened_index")) {
+		return _FileOpenAtStartup(msg);
+	}
 
-	status_t status = B_OK;
+	int32 firstAdded = -1;
 	entry_ref ref;
 	int32 refsCount = 0;
 	while (msg->FindRef("refs", refsCount++, &ref) == B_OK) {
-		// Check existence
-		BEntry entry(&ref);
-		if (!entry.Exists())
-			continue;
-		// first let's see if it's already opened.
-		const int32 openedIndex = _GetEditorIndex(&ref);
-		if (openedIndex != -1) {
-			if (openedIndex != fTabManager->SelectedTabIndex()) {
-				fTabManager->SelectTab(openedIndex, &selectTabInfo);
-			} else {
-				if (lsp_char >= 0 && be_line > -1) {
-					fTabManager->SelectedEditor()->GoToLSPPosition(be_line - 1, lsp_char);
-				} else if (be_line > -1) {
-					fTabManager->SelectedEditor()->GoToLine(be_line);
-				}
 
-				fTabManager->SelectedEditor()->GrabFocus();
-			}
-
-			// apply LSP edits
-			std::string edit = msg->GetString("edit", "");
-			if (!edit.empty())
-				fTabManager->SelectedEditor()->ApplyEdit(edit);
+		const int32 be_line   = msg->GetInt32("start:line", msg->GetInt32("be:line", -1));
+		const int32 lsp_char  = msg->GetInt32("start:character", -1);
+		const bool openWithPreferred = msg->GetBool("openWithPreferred", false);
 
-			opened_index = -1;
-			continue;
+		int32 index = _GetEditorIndex(&ref);
+		if (index != -1) {
+			_SelectEditorToPosition(index, be_line, lsp_char);
+		} else {
+			if(_FileOpenWithPosition(&ref , openWithPreferred, be_line, lsp_char) != B_OK)
+				continue;
+			index = _GetEditorIndex(&ref);
 		}
 
-		if (!_FileIsSupported(&ref)) {
-			if (openWithPreferred)
-				_FileOpenWithPreferredApp(&ref); // TODO: make this optional?
-			continue;
-		}
+		_ApplyEditsToSelectedEditor(msg);
+		if (firstAdded == -1)
+			firstAdded = index;
+	}
+
+	if (firstAdded > -1 && fTabManager->CountTabs() > firstAdded) {
+		fTabManager->SelectTab(firstAdded);
+	}
 
-		// register the file as a recent one.
-		be_roster->AddToRecentDocuments(&ref, GenioNames::kApplicationSignature);
+	return B_OK;
+}
 
-		// new file to load..
-		selectTabInfo.AddBool("caret_position", true);
-		int32 index = fTabManager->SelectedTabIndex() + 1;
-		Editor* editor = _AddEditorTab(&ref, index, &selectTabInfo);
+void
+GenioWindow::_ApplyEditsToSelectedEditor(BMessage* msg)
+{
+	// apply LSP edits
+	std::string edit = msg->GetString("edit", "");
```

Comment:
Could we use BString here? Just for consistence

Reply:
This std::string is used directly by our json library directly. We could use a BString but then anyway we have to conver to a std::stream before invoking the library.. and I'm not sure it's worth doing it.

Reply:
Ok, that makes sense